src/libraries/System.Private.CoreLib/src/System/Reflection/FieldAccessor.cs

```diff
@@ -0,0 +1,440 @@
+        private readonly FieldAccessorType _fieldAccessType;
+        private readonly PrimitiveFieldSize _primitiveFieldSize;
+
+        internal FieldAccessor(FieldInfo fieldInfo)
+        {
+            _fieldInfo = (RtFieldInfo)fieldInfo;
+            RuntimeType? declaringType = (RuntimeType?)_fieldInfo.DeclaringType;
+
+            // Call the class constructor if not already.
+            if (declaringType is null)
+            {
+                InvokeModuleConstructor(fieldInfo.Module);
+            }
+            else if (!declaringType.DomainInitialized)
+            {
+                InvokeClassConstructor(fieldInfo);
+                declaringType.DomainInitialized = true;
+            }
+
+            // Cached the method table for performance.
+            unsafe
+            {
+                _methodTable = (MethodTable*)_fieldInfo.FieldType.TypeHandle.Value;
+            }
+
+            // Initialize for the type of field.
+            if (declaringType is not null && declaringType.ContainsGenericParameters)
+            {
+                _fieldAccessType = FieldAccessorType.NoInvoke;
+            }
+            else if (!RuntimeFieldHandle.IsFastPathSupported(_fieldInfo))
+            {
+                // Currently this is true for [ThreadStatic] cases and for fields added from EnC.
+                _fieldAccessType = FieldAccessorType.SlowPath;
+            }
+            else
+            {
+                Type fieldType = _fieldInfo.FieldType;
+
+                if (fieldInfo.IsStatic)
+                {
+                    _addressOrOffset = RuntimeFieldHandle.GetStaticFieldAddress(_fieldInfo, out bool isBoxed);
+
+                    if (fieldType.IsValueType)
+                    {
+                        // The runtime stores non-primitive value types as a boxed value.
+                        _fieldAccessType = isBoxed ? FieldAccessorType.StaticValueTypeBoxed : FieldAccessorType.StaticValueType;
+                    }
+                    else if (fieldType.IsPointer)
+                    {
+                        Debug.Assert(!isBoxed);
+                        _fieldAccessType = FieldAccessorType.StaticPointerType;
+                    }
+                    else if (fieldType.IsFunctionPointer)
+                    {
+                        Debug.Assert(!isBoxed);
+                        _fieldAccessType = FieldAccessorType.StaticValueType;
+                        unsafe
+                        {
+                            _methodTable = (MethodTable*)typeof(IntPtr).TypeHandle.Value;
+                        }
+                    }
+                    else
+                    {
+                        Debug.Assert(!isBoxed);
+                        _fieldAccessType = FieldAccessorType.StaticReferenceType;
+                    }
+                }
+                else
+                {
+                    _addressOrOffset = RuntimeFieldHandle.GetInstanceFieldOffset(_fieldInfo);
```

Comment:
AFAIK fields added in EnC don't have a valid offset that's inside of the object.